Improve stability by changing the convergence check

andlaus opened the pull request:
This PR reduces the maximum number of Newton iterations from 15 to 10, and -- more importantly -- it ignores the maximum-residual-in-all-cells criterion (CNV) for the nineth and the tenth iterations, i.e., after iteration 8 only the aggregate mass balance and the well residual criterions are considered.

I've tested this on Norne, Model 2, Model 2 realization 0 and realization 5 and I made sure that all opm-simulators tests still pass on my machine. With this PR merged, the simulator seems to be significantly more stable (e.g. it works fine with OPM/opm-material#220 applied) and performance seems to also improve by 5 to 10% even for cases  where complete breakdowns do not occur, i.e. Norne.

The results of the unit tests stay identical. That said, since the time stepping is different, the results produced for Norne are slightly different from the current master version. This considered, one needs a magnifying glass to see the difference and on average the result curves are not further away from those of E100 than the current master.

andlaus commented:
jenkins build this please

alfbr commented:
Great! Looking forward to testing it Monday. Maybe it helps for the fast ILU patch too.

joakim-hove commented:
I guess this behaviour should be configurable?

alfbr commented:
Not necessarily (given it is a uniform improvement).

akva2 commented:
ehrm, that failure is my fault. it needs a explicit RestartKey around the enum value.

andlaus commented:
so far I haven't seen any regression, but I don't have a cluster to test it on whole ensembles.

akva2 commented:
jenkins build this please

atgeirr commented:
I like the idea! Something a little fishy about the duplicated max_iter_s though? Must take a look in broad daylight... (on Monday).

blattms commented:
Out of curiosity:
1. How often do we converge in less than 8 nonlinear iterations?
2. What happens if we drop the local convergence check CNV altogether?

blattms commented:
And shouldn't this change also be applied elsewhere?

andlaus commented:
> How often do we converge in less than 8 nonlinear iterations?

if nothing goes wrong (i.e., Norne), my gut feeling is that about 80% of all substeps do.

>  What happens if we drop the local convergence check CNV altogether?

the results will change significantly. Not for Norne or the SPEs, but e.g. for model 2. That said, @totto82 brought up the idea of just disabling the incriminating cells for the CNV criterion after they hit a time step where the CNV value is more than the tolerance after more than 8 iterations. I think this is a very good idea, but it is quite a bit more elaborate to implement and it comes with some headaches for restart so I decided to initially do it as proposed in this PR.

andlaus commented:
> And shouldn't this change also be applied elsewhere?

which places do you have in mind?

blattms commented:
flow_legacy, solvent, polymer

andlaus commented:
> flow_legacy, solvent, polymer

I guess this should not be done for them because nowadays the main point of these simulators is to have a stable base to compare with, i.e. regressions on them should be avoided at all costs? (note that @totto82 is currently working on replacements flow_{solvent,polymer}. the solvent part is already merged, but currently not exposed without explicitly enabling it in the source code, cf. #1176)

andlaus commented:
@alfbr: did you hit on any reason not to merge this?

alfbr commented:
No, this is good to go from my end. No full ensemble testing, but that cannot be expected for every PR anyway.

andlaus commented:
cool! lacking further objections I'll merge this tomorrow.

atgeirr commented:
I am in favour of merging the *behaviour* but not quite the implementation. It duplicates max_iter across two different classes, blurring their responsibilities and placing defaults (that should be identical) in two different places, making it possible to mess it up in the future. Also, the behaviour should be user-controllable, for example even if we change the default behaviour it should be possible to get the old behaviour (at least for something simple like this). My suggestion is:
 - Do *not* add "max_iter" to BlackoilModelParameters, instead add "strict_iter" (or similar name) which says for how many iterations one should require strict behaviour. Then check against that instead of max_iter - 2.

andlaus commented:
> I am in favour of merging the behaviour but not quite the implementation. It duplicates max_iter across two different classes, blurring their responsibilities and placing defaults (that should be identical)

I also noticed this. Since this is very easy to get wrong, so I consider it an issue with design of the current parameter system. Anyway, I've implemented the changes you've suggested.

atgeirr commented:
jenkins run norne please

blattms commented:
Never did take a look at the SPE5 test case. Is getting
```
[NON-XML-CHAR-0x1B][31;1mBug: NaN residual for phase Oil[NON-XML-CHAR-0x1B][0m
[NON-XML-CHAR-0x1B][35;1mProblem: [/var/lib/jenkins/workspace/opm-simulators-PR-builder/opm/autodiff/BlackoilModelBase_impl.hpp:1834] NaN residual for phase Oil
```
expected?

akva2 commented:
i cannot comment on the expectedness in general, but that is not caused by this pr, it is triggered in the 'original' code as well.

akva2 commented:
however, it causes new behavior after the chopping.
old code:
```
Time step   36 at day 1095/8034, date = 31-Dec-1989
  Substep 0, stepsize 31 days.
    Switching control mode for well PROD from SURFACE_RATE to BHP
    Switching control mode for well PROD from BHP to SURFACE_RATE
    Switching control mode for well INJG from SURFACE_RATE to BHP
Bug: NaN residual for phase Oil[0m
Problem: Solver convergence failure - Numerical problem encountered
         Timestep chopped to 10.230000 days
  Substep 0, stepsize 10.23 days.
    Switching control mode for well PROD from SURFACE_RATE to BHP
    Substep summary: well its =  2, newton its =  6, linearizations =  7 ( 0.389 sec), linear its =  20 ( 0.152 sec)
  Substep 1, stepsize 20.77 days.
    Substep summary: newton its = 12, linearizations = 13 ( 0.837 sec), linear its =  46 ( 0.364 sec)
```
new code:
```
Time step   36 at day 1095/8034, date = 31-Dec-1989
  Substep 0, stepsize 31 days.
    Switching control mode for well PROD from SURFACE_RATE to BHP
    Switching control mode for well PROD from BHP to SURFACE_RATE
    Switching control mode for well INJG from SURFACE_RATE to BHP
Bug: NaN residual for phase Oil[0m
Problem: [/var/lib/jenkins/workspace/opm-simulators-PR-builder/opm/autodiff/BlackoilModelBase_impl.hpp:1834] NaN residual for phase Oil
         Timestep chopped to 10.230000 days.
  Substep 0, stepsize 10.23 days.
    Switching control mode for well PROD from SURFACE_RATE to BHP
    Substep summary: well its =  2, newton its =  6, linearizations =  7 ( 0.106 sec), linear its =  20 ( 0.012 sec)
  Substep 1, stepsize 20.77 days.
Problem: [/var/lib/jenkins/workspace/opm-simulators-PR-builder/opm/autodiff/NonlinearSolver_impl.hpp:166] Solver convergence failure - Failed to complete a time step within 10 iterations.
         Timestep chopped to 6.854100 days.
  Substep 1, stepsize 6.8541 days.
    Substep summary: newton its =  3, linearizations =  4 ( 0.042 sec), linear its =   8 ( 0.004 sec)
  Substep 2, stepsize 13.9159 days.
    Substep summary: newton its =  4, linearizations =  5 ( 0.052 sec), linear its =  13 ( 0.005 sec)
```

akva2 commented:
as an alternative to updating the files you can use https://github.com/OPM/opm-data/pull/226 and https://github.com/akva2/opm-simulators/commit/2a1c7537070a0af5af36919e23de4492a828428f.

if @andlaus can rebase and then pick the commit on top we can test it (worked fine on my box). the commit for this pr contains some generic fixes, so if you do not go this route, i will split those out. 

a polymer update is needed as the bug led to the param file not being used. i worked around this right now, but it should be done at some point.

andlaus commented:
> as you can see the old code did 12 newton its. as 12 > 10 different behavior will be triggered.

this should be expected in some cases. what I've seen so far is that it does not converge in the vast majority of the cases which did not converge after 10 iterations. I guess the best way to do timestep control would be to ask a crystal ball ;)

> as an alternative to updating the files you can use OPM/opm-data#226 and akva2/opm-simulators@2a1c753.

cool. I've cherry picked akva2/opm-simulators@2a1c753 and rebased on master. (the message() in 2a1c753 seemed to be a debugging artiact, so I removed it.)

andlaus commented:
jenkins build this opm-data=226 please

akva2 commented:
yes, nothing unexpected going on, i just explained it :) and yes,  leaving the message was an error.

andlaus commented:
jenkins seems like to have stumbled on the last attempt: it says it merged the HEAD, but there's no build for this in the queue and there are idle builders, strange. let's try again:

andlaus commented:
jenkins build this opm-data=226 please

akva2 commented:
it's only one job per module due to the workspace. it's busy running norne..

andlaus commented:
ah, okay. I didn't expect that, what's the reason for this?

akva2 commented:
as i said it's because of the workspace. it's one workspace per job, and we cannot build two things in the same place. i could setup multiple/dynamic workspaces but it would significantly increase network traffic and i have been told to limit this.

andlaus commented:
I thought that each workspace was independent of the others so there could be two of them testing the same module. Anyway, it's not a big deal.

On other news, in my machine Norne works fine the current version of this PR and all unit tests pass if OPM/opm-data is applied. the failing SPE5 test on the opm-data master was added after this PR was opened, right? The Norne performance still seems to be 5 to 10% better...

akva2 commented:
yes the test was added after you opened this so you would be blissfully unaware ;) the workspaces doubles as jenkins' git clone, so if there were to be two opm-simulators jobs running simultaneously, i would have to clone the module fresh for each build (hence the additional network traffic) and handle the git stuff myself). not a biggie but some additional code that can bug up;))

andlaus commented:
wouldn't it be equally simple to just clone a local master version of all modules, then fetch the PR from the github remote and merge this into the cloned master? just wondering, but as said: not a big deal.

alfbr commented:
Is this one ready now?

andlaus commented:
from my perspective, yes.

@atgeirr: would you Press Green? While at it, OPM/opm-material#220 should also be ready after the present PR has been merged.

atgeirr commented:
Yes, I will merge!
